File: Final/erp-backend/src/main/java/com/pyme/erp/service/ProductoService.java
package com.pyme.erp.service;

import com.pyme.erp.dto.request.ProductoRequest;
import com.pyme.erp.model.Producto;

import java.util.List;

public interface ProductoService {
    
    Producto crear(ProductoRequest request);
    
    Producto actualizar(Long id, ProductoRequest request);
    
    Producto obtenerPorId(Long id);
    
    List<Producto> listarActivos();
    
    void desactivar(Long id);
    
    void activar(Long id);
    
    List<Producto> buscarPorNombre(String nombre);
    
    List<Producto> obtenerProductosConStockBajo();
    
    void actualizarStock(Long productoId, Integer cantidad, String operacion);
}

File: Final/erp-frontend/erp-frontend/src/app/services/producto.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ResponseWrapper, Producto } from '../models/response-wrapper.model';

@Injectable({
  providedIn: 'root'
})
export class ProductoService {
  private baseUrl = 'http://localhost:8080/api/productos';

  constructor(private http: HttpClient) { }

  crear(producto: Producto): Observable<ResponseWrapper<Producto>> {
    return this.http.post<ResponseWrapper<Producto>>(this.baseUrl, producto);
  }

  listar(): Observable<ResponseWrapper<Producto[]>> {
    return this.http.get<ResponseWrapper<Producto[]>>(this.baseUrl);
  }

  obtenerPorId(id: number): Observable<ResponseWrapper<Producto>> {
    return this.http.get<ResponseWrapper<Producto>>(`${this.baseUrl}/${id}`);
  }

  actualizar(id: number, producto: Producto): Observable<ResponseWrapper<Producto>> {
    return this.http.put<ResponseWrapper<Producto>>(`${this.baseUrl}/${id}`, producto);
  }

  desactivar(id: number): Observable<ResponseWrapper<void>> {
    return this.http.delete<ResponseWrapper<void>>(`${this.baseUrl}/${id}`);
  }

  obtenerStockBajo(): Observable<ResponseWrapper<Producto[]>> {
    return this.http.get<ResponseWrapper<Producto[]>>(`${this.baseUrl}/stock-bajo`);
  }

  actualizarStock(id: number, cantidad: number, operacion: string): Observable<ResponseWrapper<void>> {
    return this.http.put<ResponseWrapper<void>>(`${this.baseUrl}/${id}/stock?cantidad=${cantidad}&operacion=${operacion}`, {});
  }
}